### PR TITLE
Hello Sparcc

### PR DESCRIPTION
**Pull request by d-callan**
adding support for sparcc courtesy SpiecEasi.. its a private repo though, which i dont love. the only other option i see is maintaining a sparcc implementation ourselves though, which we can always switch to if needed in the future.

### CODE REVIEW

File: R/method-correlation.R [hunk 1]
```diff
@@ -136,11 +145,29 @@ setMethod("correlation", signature("data.table", "missing"), function(data1, dat
...
-  # na.or.complete removes rows with NAs, if no rows remain then correlation is NA
   # keep matrix for now so we can use lower.tri later, expand.grid will give us the needed data.frame
-  corrResult <- Hmisc::rcorr(as.matrix(data1), type = method)
-  pVals <- corrResult$P
-  corrResult <- corrResult$r
+  if (method == 'sparcc') {
+
+    # get the sparcc function from the SpiecEasi namespace and make a local copy that do.call can find
+    sparcc <- get("sparcc", asNamespace("SpiecEasi"))
+    # sub-sampled `statistic` functions to pass to `boot::boot` that we can use to find pvalues
```
Comment (asizemore) [inline R/method-correlation.R:153]: Okay i found why this is hapening, but i had to go to the sparcc implementation to understand what, why, and how. Mind adding some comments to explain what we're doing here please?

File: R/method-correlation.R [hunk 2]
```diff
@@ -136,11 +145,29 @@ setMethod("correlation", signature("data.table", "missing"), function(data1, dat
...
+    statisticperm=function(data, indices) do.call("sparcc", c(list(apply(data[indices,], 2, sample)), list()))$Cor
+    statisticboot=function(data, indices) do.call("sparcc", c(list(data[indices,,drop=FALSE]), list()))$Cor
+
+    # calling the bootstrap version of sparcc and finding pvalues
+    corrResult <- SpiecEasi::pval.sparccboot(SpiecEasi::sparccboot(data1, statisticboot = statisticboot, statisticperm = statisticperm, R = 100))
```
Comment (asizemore) [inline R/method-correlation.R:158]: Can we name this something different? It's confusing when variables get redefined later

Comment (asizemore) [reply R/method-correlation.R:158]: looking below i see we're already in a pattern of doing this lol

File: R/method-correlation.R [hunk 3]
```diff
@@ -298,12 +325,12 @@ setMethod("selfCorrelation", signature("SampleMetadata"), function(data, method
...
-#' @param method string defining the type of correlation to run. The currently supported values are 'spearman' and 'pearson'
+#' @param method string defining the type of correlation to run. The currently supported values are 'spearman', 'pearson' and 'sparcc'
 #' @param verbose boolean indicating if timed logging is desired
 #' @return ComputeResult object
 #' @import veupathUtils
 #' @export
-setMethod("selfCorrelation", signature("data.table"), function(data, method = c('spearman','pearson'), verbose = c(TRUE, FALSE)) {
+setMethod("selfCorrelation", signature("data.table"), function(data, method = c('spearman','pearson','sparcc'), verbose = c(TRUE, FALSE)) {
   correlation(data, method=method, verbose=verbose)
 })
```
Comment (asizemore) [inline R/method-correlation.R:336]: Why not sparcc as a method option in line 343? We could be correlating two compositional assays there

File: tests/testthat/test-correlation.R [hunk 4]
```diff
@@ -39,9 +39,9 @@ test_that('correlation works with two data tables', {
 test_that("correlation works with a single data table", {
   nSamples = 200
   testData1 <- data.table(
-    "contA1" = rnorm(nSamples),
-    "contB1" = rnorm(nSamples),
-    "contC1" = rnorm(nSamples)
+    "contA1" = rnorm(nSamples,1,.5), #sparcc wont like negative values
```
Comment (asizemore) [inline tests/testthat/test-correlation.R:42]: We'd cover more cases if we left this as is and then sent an absolute value version of the data to tests with sparcc. Don't want to accidentally make a change later that breaks using negative values in the non-sparcc case!

File: tests/testthat/test-correlation.R [hunk 5]
```diff
@@ -237,6 +242,16 @@ test_that("correlation returns an appropriately structured result for assay agai
   #expect_equal(as.character(unique(statsData$data1)), names(testOTU)[2:(length(names(testOTU)) - 1)])
   #expect_equal(as.character(unique(statsData$data2)), names(testOTU)[3:length(names(testOTU))])
   expect_true(all(!is.na(statsData)))
+
+  # method = sparcc
+  result <- selfCorrelation(data, method='sparcc', verbose = FALSE)
```
Comment (asizemore) [inline tests/testthat/test-correlation.R:247]: this should fail with the negative data passed to it, no?

File: tests/testthat/test-correlation.R [hunk 6]
```diff
@@ -237,6 +242,16 @@ test_that("correlation returns an appropriately structured result for assay agai
...
+  statsData <- result@statistics@statistics
+  expect_s3_class(statsData, 'data.frame')
+  expect_equal(names(statsData), c('data1','data2','correlationCoef','pValue'))
+  expect_equal(nrow(statsData), 66) # Should be number of taxa * number of taxa
+  #expect_equal(as.character(unique(statsData$data1)), names(testOTU)[2:(length(names(testOTU)) - 1)])
```
Comment (asizemore) [inline tests/testthat/test-correlation.R:252]: Can we delete these or reimplement the tests? Not sure why they would be failing

File: R/method-correlation.R (hunk 3 above)
Comment (d-callan) [reply R/method-correlation.R:336]: yea ive thought about it a fair bit actually.. im not sure the compositional adjustment is appropriate if youre correlating across two different compositions.

File: tests/testthat/test-correlation.R (hunk 5 above)
Comment (d-callan) [reply tests/testthat/test-correlation.R:247]: what negative data? df is our testOTU here i think.

File: tests/testthat/test-correlation.R (hunk 6 above)
Comment (d-callan) [reply tests/testthat/test-correlation.R:252]: if we prefilter, we no longer know which features will make it into the result. i suppose i could explicitly set all the prefilters to 0 here though

File: R/method-correlation.R [hunk 7]
```diff
@@ -136,11 +145,29 @@ setMethod("correlation", signature("data.table", "missing"), function(data1, dat
...
   ## Compute correlation
   # rownames and colnames should be the same in this case
-  # na.or.complete removes rows with NAs, if no rows remain then correlation is NA
   # keep matrix for now so we can use lower.tri later, expand.grid will give us the needed data.frame
-  corrResult <- Hmisc::rcorr(as.matrix(data1), type = method)
-  pVals <- corrResult$P
-  corrResult <- corrResult$r
+  if (method == 'sparcc') {
+
+    # get the sparcc function from the SpiecEasi namespace and make a local copy that do.call can find
```
[the remark this answers is not in the document]
Comment (d-callan) [reply R/method-correlation.R:151]: i guess we dont have too.. but if we ever want to be able to pass args to sparcc well have to bring it back. ill leave it commented w a note.

**Pull request merged**